Network stack boot time optimization: decouple from DATA partition

ytsssun opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

  Closes #637

**Description of changes:**

This PR optimizes boot time by decoupling the network stack initialization from the DATA partition, allowing network services to start in parallel with local filesystem operations.

### Boot sequence change overview
| Before | After |
|--------|-------|
| <img width="600" alt="image(1)" src="https://github.com/user-attachments/assets/f0d77349-df69-4f82-ba5f-a6506fe8be8e" /> | <img width="600" alt="After" src="https://github.com/user-attachments/assets/2fd04ce9-2244-46f6-b22e-f97284debfc6" /> |



### Direct PRIVATE Partition Mount
- Added direct mount point `/.bottlerocket` for the PRIVATE partition that bypasses dependency on DATA partition
- Created a bind mount from `/.bottlerocket` to `/var/lib/bottlerocket` for backward compatibility
- Network services now use `/.bottlerocket/netdog` directly instead of `/var/lib/netdog`

### Netdog Updates
- Updated path constants to use `/.bottlerocket/netdog` instead of `/var/lib/netdog`
- Drop default dependencies and update `generate-network-config.service` with /.bottlerocket dependencies.
- Create `/etc/sysctl.d` with `ExecStart` entry for `write-network-status.service`.
- Drop default dependencies for `write-network-status.service` and explicitly add `Before=network-online.target` for `write-network-status.service`.
- Removed dependency on `systemd-tmpfiles` for directory creation

### D-Bus and Network Service Improvements
- Added `DefaultDependencies=no` to `dbus.socket` to allow early D-Bus initialization
- Made `network-pre.target` require `dbus-broker.service` to ensure D-Bus is ready before network services start
- Disabled `PrivateTmp` for `dbus-broker.service` to remove dependency on `systemd-tmpfiles` and `local-fs.target`


**Testing done:**

<details>
  <summary>Click to expand testing details</summary>

  ### Functional Testing
  - Tested using `metal-dev` instance to confirm that `net.toml` injected to PRIVATE partition can still be read and that
  `write-network-status.service` can write to the `/.bottlerocket/netdog` path with no issue
  - Verified the node booted properly with all network functionality intact
  - Used systemd-analyze plot to generate the boot plot and confirmed the decoupling of the network stack and DATA
  partition
  - Verified that network initialization and local-fs (DATA) now occur in parallel
  - Confirmed network configuration and state files are properly persisted across reboots
  - Validated that all existing network functionality is maintained with the new directory structure

  ### Boot-time Testing
  - Initial tests on m5.xlarge instances show ~987ms improvement in network-online.target timing and ~817ms improvement in configured.target timing
  - Ran over 10k sample boots to measure the improvement and we are seeing about 700ms improvement statistically.


  </details>

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

ytsssun commented:
Pushed new commit to address a race condition we noticed during our test:

Networkd service started too early and it sometimes failed to get the status info of the network interface from D-Bus because D-Bus was not yet ready. 

Added change to initialize D-Bus earlier and enforced dependency on dbus-broker.service in network-pre.target.

bcressey commented on packages/netdog/generate-network-config.service:
```diff
@@ -1,9 +1,14 @@
 RefuseManualStop=true
+Before=network-pre.target
+After=\x2ebottlerocket-netdog.mount
+Requires=\x2ebottlerocket-netdog.mount
```
These are redundant with `RequiresMountsFor`:
```suggestion
After=\x2ebottlerocket-netdog.mount
Requires=\x2ebottlerocket-netdog.mount
```

bcressey commented on packages/release/var-lib-bottlerocket.mount:
```diff
@@ -2,14 +2,15 @@
 Conflicts=umount.target
-RequiresMountsFor=/var
+RequiresMountsFor=/var /.bottlerocket
+After=\x2ebottlerocket.mount
```
implied by `RequiresMountsFor`:
```suggestion
```

bcressey commented on packages/release/var-lib-bottlerocket.mount:
```diff
@@ -2,14 +2,15 @@
-Type=ext4
-Options=defaults,nosuid,nodev,noexec,noatime,private,context=system_u:object_r:private_t:s0
+Type=none
+Options=bind,nosuid,nodev,noexec,noatime,private,context=system_u:object_r:private_t:s0
```
Most of these options aren't valid for bind mounts because they share the same filesystem superblock.

Generally it's useful to set up the recursive options so mounts are propagated in both locations:
```suggestion
Options=rbind,rshared
```

bcressey commented on packages/selinux-policy/fs.cil:
```diff
@@ -83,6 +83,8 @@
 (filecon "/var/lib/netdog" any lease)
 (filecon "/var/lib/netdog/.*" any lease)
+(filecon "/.bottlerocket/netdog" any local)
+(filecon "/.bottlerocket/netdog/.*" any local)
```
These `filecon` directives won't have any effect unless you're either creating files here at build time, or relabeling files at runtime. You can remove them.

bcressey commented on packages/selinux-policy/rules.cil:
```diff
@@ -151,6 +151,7 @@
 ; processes, so those credentials need to grant a superset of permissions.
 ; This lets us further restrict what other host processes can execute.
 (allow mount_t global (files (execute)))
+(allow mount_t private_t (files (mutate)))
```
This rule is being added in a place where the comment doesn't apply - it doesn't have anything to do with executing programs.

Let's add `mount_t` to `api_s` in `subject.cil` instead.

ytsssun replied:
The `mount_t` is no longer needed now that we grant netdog directly `api_exec_t` access.

bcressey commented on packages/release/bottlerocket.mount.in:
```diff
@@ -0,0 +1,19 @@
+What=/dev/disk/by-partlabel/BOTTLEROCKET-PRIVATE
+Where=/.bottlerocket
+# "noexec" omitted to allow netdog to run
+Options=defaults,nosuid,nodev,noatime,private,context=system_u:object_r:private_t:s0
```
Any time these are both true:
1. You are applying a `context=` mount option.
2. You are in a unit with `DefaultDependencies=no`.

Then you must add these dependencies:
```
After=selinux-policy-files.service
Wants=selinux-policy-files.service
```

Otherwise the context option can be silently dropped, if `mount` doesn't see the policy files and concludes that SELinux is not actually enabled.

bcressey commented on packages/release/bottlerocket.mount.in:
```diff
@@ -0,0 +1,19 @@
+[Mount]
+What=/dev/disk/by-partlabel/BOTTLEROCKET-PRIVATE
+Where=/.bottlerocket
+# "noexec" omitted to allow netdog to run
```
Do not omit the `noexec` option, `netdog` is not an executable on this filesystem.

bcressey commented on packages/release/bottlerocket.mount.in:
```diff
@@ -0,0 +1,19 @@
+[Unit]
+Description=Mount Bottlerocket persistent data directory (/.bottlerocket)
```
Please follow the existing conventions for descriptions in all the mount units you're adding:
```suggestion
Description=Private Directory (./bottlerocket)
```

Check `grep Description packages/*/*.mount` for more examples.

bcressey commented on packages/release/bottlerocket-netdog.mount.in:
```diff
@@ -0,0 +1,17 @@
+Conflicts=umount.target
+Before=local-fs.target umount.target
+RequiresMountsFor=/.bottlerocket
+DefaultDependencies=no
```
nit: I would list `DefaultDependencies` first since it's important for the reader's understanding of whether the rest of the dependencies are complete.

bcressey commented on packages/release/bottlerocket-netdog.mount.in:
```diff
@@ -0,0 +1,17 @@
+[Unit]
```
Same notes for this unit re: description, dependencies on SELinux policy files.

(Technically the SELinux policy dependency will already be satisfied by the `/.bottlerocket` mount. But the more mount units that combine `DefaultDependencies=no` with Wants/After on SELinux policy files, the better - that way copy/pasted units are likely to be correct instead of subtly broken.)

ytsssun replied:
We no longer need this unit as we are not longer using overlayfs approach.

bcressey commented on packages/release/bottlerocket-netdog.mount.in:
```diff
@@ -0,0 +1,17 @@
+What=overlay
+Where=/.bottlerocket/netdog
+Type=overlay
+Options=lowerdir=/.bottlerocket/.netdog/lower,upperdir=/.bottlerocket/.netdog/upper,workdir=/.bottlerocket/.netdog/work,context=system_u:object_r:local_t:s0
```
Why can't we use `lease_t` for this?

Effectively this weakens the older permissions for `/var/lib/netdog` from `lease_t` (restricted) to `local_t` (all containers can write.

bcressey commented on packages/release/prepare-bottlerocket-netdog.service:
```diff
@@ -0,0 +1,20 @@
+Description=Prepare Netdog OverlayFS Directories in /.bottlerocket
+DefaultDependencies=no
+After=selinux-policy-files.service
+Wants=selinux-policy-files.service
```
You don't need these here since they should be dependencies of the `.bottlerocket.mount`.

bcressey commented on packages/release/prepare-bottlerocket-netdog.service:
```diff
@@ -0,0 +1,20 @@
+[Unit]
+Description=Prepare Netdog OverlayFS Directories in /.bottlerocket
```
For consistency with the other combinations of "mount units" and "helpers that create some directories after the mount":
```suggestion
Description=Prepare Private Directory (/.bottlerocket)
```

I would also rename the service to `prepare-bottlerocket.service` since we might add other preparations later.

ytsssun replied:
I updated to be `Prepare Netdog Directory (/.bottlerocket/netdog)`

Let me know if that makes sense.

bcressey commented on packages/release/prepare-bottlerocket-netdog.service:
```diff
@@ -0,0 +1,20 @@
+StandardError=journal+console
+
+# /.bottlerocket/netdog will be the mount point.
+# /.bottlerocket/.netdog/{lower,upper,work} are the backing directories.
```
I don't think this makes sense here out of context. It might be better to add as comments in the subsequent mount unit to explain why the "prepare" dependency exists.
```suggestion
```

bcressey commented on sources/api/storewolf/src/main.rs:
```diff
@@ -518,6 +576,10 @@ fn usage() -> ! {
+            [ subcommand ]
+
+        Subcommands:
+            prepare-netdog    Prepare the netdog directories for overlayfs mount
```
This should probably get migrated to `argh` before it gets much more complicated. Then we can add a subcommand for `prepare-datastore` to handle the existing logic.

`prepare-netdog-dirs` might be more clear than `prepare-netdog`.

bcressey commented on sources/api/storewolf/src/main.rs:
```diff
@@ -499,13 +507,63 @@ where
+///    - /.bottlerocket/.netdog/lower
+///    - /.bottlerocket/.netdog/upper
+///    - /.bottlerocket/.netdog/work
+///    - /.bottlerocket/netdog
```
Somewhere we should explain why this is actually required:
1. The filesystem is mounted with the `private_t` that prevents directories under it from having a different label.
2. We don't want to allow `netdog` to write to `private_t`, so we're setting up an overlayfs with a label that it can write to.

One thing to think about now that `wicked` is gone - we no longer have the case where `netdog` is invoked by a confined, unprivileged daemon as a DHCP helper. It's all systemd invocations by one-shot units.

So it might be simpler overall to give `netdog` the same treatment as `storewolf` in `fs.cil` - make it an `api_exec` program so that when invoked by systemd, it's able to write to `private_t` files.

Then we wouldn't need the overlayfs mount or the storewolf changes.

ytsssun replied:
Moved the subcommand under netdog.

ytsssun committed:
release: add direct mount for PRIVATE partition at /.bottlerocket

Create a direct mount point (/.bottlerocket) for the PRIVATE partition that
bypasses the dependency on the DATA partition. Previously, /var/lib/bottlerocket
was backed by the PRIVATE partition but required DATA partition availability
through the /var -> /local/var bind mount chain.

This change maintains backward compatibility by adding a bind mount from
/.bottlerocket to /var/lib/bottlerocket, allowing network services to
initialize in parallel with local-fs operations.

Signed-off-by: Yutong Sun <yutongsu@amazon.com>

ytsssun committed:
systemd: ensure dbus is available for network services

Fix race condition where network services attempt to use D-Bus functionality
before the D-Bus broker is fully initialized:

1. Add DefaultDependencies=no to dbus.socket to allow D-Bus to start earlier
   in the boot process
2. Make network-pre.target require dbus.service to ensure D-Bus is
   ready before network services start
3. Disable PrivateTmp for dbus-broker.service to remove dependency on
   systemd-tmpfiles and local-fs.target
4. Apply the same changes to whipet.service as whipet.service is also aliased
as dbus.service

Signed-off-by: Yutong Sun <yutongsu@amazon.com>

ytsssun committed:
migrator: add explicit dependency on /var/lib/bottlerocket

Migrator depends on /var/lib/bottlerocket/, previously it was the mount point
to PRIVATE partition, now that /.bottlerocket is the mount point and /var/lib/bottlerocket
became the bind mount, the explicit dependency would be necessary.

Signed-off-by: Yutong Sun <yutongsu@amazon.com>

ytsssun commented:
Synced with @bcressey , we are adopting different implementation here - instead of using overlayfs for the /.bottlerocket/netdog to grant netdog access to the files on private partition, we will just grant netdog `api_exec_t` to allow access to `private_t`. 

Pushed code changes for that.

ytsssun commented:
Pushed a minor change to add `/etc/sysctl.d` as `ReadWritePath` to `write-network-status.service`, otherwise will get error like
```
Failed to write sysctl config to '/etc/sysctl.d/90-primary_interface.conf': Read-only file system (os error 30)
```

bcressey commented on sources/netdog/src/main.rs:
```diff
@@ -12,7 +12,7 @@ It contains two subcommands meant for use as settings generators:
 
 The subcommand `generate-net-config` generates the network interface configuration for the host. If
-a `net.toml` file exists in `/var/lib/bottlerocket`, it is used to generate the configuration. If
+a `net.toml` file exists in `/.bottlerocket/netdog`, it is used to generate the configuration. If
```
```suggestion
a `net.toml` file exists in `/.bottlerocket`, it is used to generate the configuration. If
```

bcressey commented on packages/netdog/write-network-status.service:
```diff
@@ -16,5 +16,12 @@ ExecStart=/usr/bin/netdog write-primary-interface-status
 RemainAfterExit=true
 StandardError=journal+console
 
+# Security: Mount /usr and /boot read-only (but keep /etc accessible)
```
nit: since "accessible" is ambiguous:
```suggestion
# Security: Mount /usr, /boot, and /etc read-only
```

bcressey commented on packages/release/release.spec:
```diff
@@ -62,6 +62,7 @@ Source1026: media-cdrom.mount
 Source1027: root-.aws.mount
 Source1028: opt-cni.mount
 Source1029: opt-csi.mount
+Source1030: bottlerocket.mount.in
```
There's a separate block starting at Source1080 for "Mounts that require build-time edits" - and then the `systemd-escape` logic could be moved

ytsssun replied:
Done

ytsssun committed:
netdog: use /run/netdog for network status files

- Update path constants to use /run/netdog instead of /var/lib/netdog.
- Remove OVERRIDE_NET_CONFIG_FILE as it's no longer needed.

Signed-off-by: Yutong Sun <yutongsu@amazon.com>

ytsssun committed:
netdog: update services for /.bottlerocket path

As we introduced /.bottlerocket as the mount point to PRIVATE partition, we need to migrate the old
usage /var/lib/bottlerocket to point to /.bottlerocket so that netdog service will not wait for the
DATA partition

- Drop dependency on tmpfiles
- Drop default dependencies and update generate-network-config.service with /.bottlerocket dependencies
- Create /etc/sysctl.d with ExecStart entry for write-network-status.service
- Drop default dependencies for write-network-status.service
- Explicitly add Before=network-online.target for write-network-status.service

Signed-off-by: Yutong Sun <yutongsu@amazon.com>

ytsssun commented:
Pushed change so that we use `/run/netdog` for the network status files. Netdog already is properly configured to have access to the `/run/netdog` path, and we remove the status files on each boot, which is the same lifetime as the `/run/netdog`.

arnaldo2792 commented on packages/release/bottlerocket.mount.in:
```diff
@@ -0,0 +1,19 @@
+Before=local-fs.target umount.target
+After=dev-disk-by\x2dpartlabel-BOTTLEROCKET\x2dPRIVATE.device selinux-policy-files.service
+Requires=dev-disk-by\x2dpartlabel-BOTTLEROCKET\x2dPRIVATE.device
+Wants=selinux-policy-files.service
```
Why not making this `Required` as well?

ytsssun replied:
I used ` Wants` here to be consistent with elsewhere https://github.com/search?q=repo%3Abottlerocket-os%2Fbottlerocket-core-kit+%3Dselinux-policy-files.service&type=code

arnaldo2792 commented on packages/release/release.spec:
```diff
@@ -301,6 +302,11 @@ LIBEXECDIRPATH=$(systemd-escape --path %{_cross_libexecdir})
+# Process bottlerocket mount template files with proper systemd naming
+BOTTLEROCKET_PATH=$(systemd-escape --path /.bottlerocket)
+cp %{S:1086} ${BOTTLEROCKET_PATH}.mount
+install -p -m 0644 ${BOTTLEROCKET_PATH}.mount %{buildroot}%{_cross_unitdir}
```
You could just do:

```suggestion
install -p -m 0644 %{S:1086} %{buildroot}%{_cross_unitdir}/${BOTTLEROCKET_PATH}.mount
```

arnaldo2792 commented on packages/release/var-lib-bottlerocket.mount:
```diff
@@ -2,14 +2,14 @@
-Type=ext4
-Options=defaults,nosuid,nodev,noexec,noatime,private,context=system_u:object_r:private_t:s0
+Type=none
+Options=rbind,rshared
```
Do you have to make this `rshared`? Why not just `rprivate`?

ytsssun replied:
Linking @bcressey 's comments - https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/638#discussion_r2322880547

>Most of these options aren't valid for bind mounts because they share the same filesystem superblock.
>Generally it's useful to set up the recursive options so mounts are propagated in both locations

bcressey commented on packages/netdog/write-network-status.service:
```diff
@@ -1,14 +1,16 @@
-Wants=systemd-networkd-wait-online.service systemd-resolved.service
+RequiresMountsFor=/.bottlerocket
+After=systemd-networkd-wait-online.service systemd-resolved.service run-netdog.mount prepare-sysctl.service
+Requires=systemd-networkd-wait-online.service systemd-resolved.service run-netdog.mount prepare-sysctl.service
```
To avoid the change to `run-netdog.mount`:
```suggestion
RequiresMountsFor=/.bottlerocket /run/netdog
After=systemd-networkd-wait-online.service systemd-resolved.service prepare-sysctl.service
Requires=systemd-networkd-wait-online.service systemd-resolved.service prepare-sysctl.service
```

I would probably also just add the `mkdir -p /etc/sysctl.d` as another ExecStart command, vs. adding another unit that systemd has to do the bookkeeping for.
The `run-netdog.mount` change is a little odd because that unit is already `WantedBy` `network-pre.target`, which is reached before any of these

ytsssun replied:
I think systemd ordering phase, `/run/netdog` gets mounted pretty early. This is just to **explicitly** call out the dependency here since we now use /run/netdog for the status files. IMO it does not hurt. Thoughts?

bcressey commented on packages/netdog/write-network-status.service:
```diff
@@ -18,3 +20,4 @@ StandardError=journal+console
 
 [Install]
 WantedBy=network-online.target
+RequiredBy=network-online.target
```
We should set either one or the other, `WantedBy` or `RequiredBy`, but not both.

arnaldo2792 replied:
You should just keep one of them.

arnaldo2792 commented on packages/netdog/write-network-status.service:
```diff
@@ -1,14 +1,16 @@
-Wants=systemd-networkd-wait-online.service systemd-resolved.service
+RequiresMountsFor=/.bottlerocket
+After=systemd-networkd-wait-online.service systemd-resolved.service run-netdog.mount prepare-sysctl.service
+Requires=systemd-networkd-wait-online.service systemd-resolved.service run-netdog.mount prepare-sysctl.service
```
You could list `/run/netdog` as a required mount

bcressey commented on packages/netdog/run-netdog.mount:
```diff
@@ -14,3 +14,4 @@ Options=nosuid,nodev,noexec,noatime,context=system_u:object_r:lease_t:s0,mode=07
 
 [Install]
 WantedBy=network-pre.target
+RequiredBy=write-network-status.service
```
```suggestion
```

bcressey commented on packages/release/release-tmpfiles.conf:
```diff
@@ -15,6 +15,5 @@ Z /var/lib/systemd 0755 root root -
 z /var/lib/systemd/random-seed 600 root root -
 R /var/lib/systemd/linger
 D /var/lib/systemd/linger 0700 root root -
-d /etc/sysctl.d 0700 root root -
```
I'm inclined to keep this for now, and just have any units that need this directory early arrange to create it for themselves.

Right now `corndog` also needs this directory, and doesn't create it, and it's conceptually weird for `corndog` to depend on `netdog` (or related units) for this.

It's outside the scope of this change, but I'd like to explore a better mechanism, like `etc.target`, where we can collect all the units needed to populate `/etc`, and we order almost everything after that, to stop the madness of every early unit needing to depend on `selinux-policy-files.service` or run the risk subtle breakage.

`prepare-sysctl.service` seems like a similar dependency where more units may end up needing it over time.

arnaldo2792 commented on packages/os/migrator.service:
```diff
@@ -1,6 +1,8 @@
 [Unit]
 Description=Bottlerocket data store migrator
+DefaultDependencies=no
```
nit on the commit message: You don't have to start with `-`

ytsssun replied:
Updated.

ytsssun commented on packages/netdog/write-network-status.service:
```diff
@@ -1,17 +1,20 @@
 RefuseManualStart=true
 RefuseManualStop=true
-Before=early-boot-config.service
+Before=early-boot-config.service network-online.target
```
I had to add this explicit `Before` annotation, otherwise I get weird systemd ordering that write-network-status happens after network-online.target:
```
network-online.target @1.979s
└─systemd-networkd-wait-online.service @1.254s +722ms
  └─systemd-networkd.service @1.189s +61ms
    └─network-pre.target @1.186s
...
```

With this change, the write-network-status.service happens before the network-online.target.
```
network-online.target @1.788s
└─write-network-status.service @1.753s +33ms      <-- listed as dependency
  └─systemd-resolved.service @2.669s +217ms
    └─systemd-sysctl.service @753ms +22ms
...
```

ytsssun commented:
Pushed change to address comments from @bcressey and @arnaldo2792

bcressey commented:
Nice!

arnaldo2792 commented on packages/dbus-broker/dbus-broker.service:
```diff
@@ -2,7 +2,9 @@
 DefaultDependencies=false
-After=dbus.socket
+# Ensure the dbus user is created before starting dbus-broker
+After=dbus.socket systemd-sysusers.service
```
You will have to make the same change to `whippet.service` now that I merged #661 :sweat_smile:

ytsssun replied:
Applied same changes to whippet.service as it is also aliased as dbus.service.

arnaldo2792 commented on packages/netdog/generate-network-config.service:
```diff
@@ -1,9 +1,12 @@
 [Unit]
```
nit: the commit message only explains what was done but not why. There is another commit in the series where you used a list and in each item, you explained why you did what you did. You should follow the same convention (or even better, favor a narrative rather than a bullet list)

arnaldo2792 commented on packages/release/network-pre-target-dbus-dep.conf:
```diff
@@ -0,0 +1,10 @@
+[Unit]
+# Ensure D-Bus is fully initialized before network services start
```
nit: missing `.` at the end of this sentence

ytsssun commented:
Force pushed changes to address comments per @arnaldo2792 
- Applied changes to `whipet.service` as it is also aliased as `dbus.service`.
- Use `dbus.service` alias in [network-pre-target-dbus-dep.conf](https://github.com/bottlerocket-os/bottlerocket-core-kit/pull/638/files#diff-0c7580c25d5612239267521339805b582a4f5255f783b26029b71cb48fed3e32) so that the drop-in can be effective to all services providing alias `dbus.service`.
